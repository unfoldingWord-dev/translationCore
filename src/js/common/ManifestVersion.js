/**
 * Nested version of the books of the bible object.
 */
export default {
  CURRENT_VERSION: 1
};
