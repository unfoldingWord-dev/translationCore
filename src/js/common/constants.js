import path from "path-extra";
import ospath from "ospath";
import packagefile from '../../../package.json';

export const APP_VERSION = packagefile.version;
export const MIN_COMPATIBLE_VERSION = packagefile.minCompatibleVersion;

export const TRANSLATION_WORDS = 'translationWords';
export const TRANSLATION_NOTES = 'translationNotes';
export const TRANSLATION_ACADEMY = 'translationAcademy';
export const TRANSLATION_HELPS = 'translationHelps';
export const WORD_ALIGNMENT = 'wordAlignment';

export const PROJECTS_PATH = path.join(ospath.home(), 'translationCore', 'projects');
export const USER_RESOURCES_PATH = path.join(ospath.home(), "translationCore", "resources");
export const THELPS_EN_RESOURCES_PATH = path.join(USER_RESOURCES_PATH, 'en', TRANSLATION_HELPS);
export const IMPORTS_PATH = path.join(ospath.home(), 'translationCore', 'imports');
export const PROJECT_INDEX_FOLDER_PATH = path.join('.apps', 'translationCore', 'index');
export const TEMP_IMPORT_PATH = path.join(ospath.home(), 'translationCore', 'imports', 'temp');

export const PROJECT_DOT_APPS_PATH = path.join('.apps', 'translationCore');
export const PROJECT_CHECKDATA_DIRECTORY = path.join(PROJECT_DOT_APPS_PATH, 'checkData');
export const STATIC_RESOURCES_PATH = path.join(__dirname, "../../../tcResources");

export const TC_VERSION = "tc_version";
export const SOURCE_CONTENT_UPDATER_MANIFEST = "source-content-updater-manifest.json";

export const tc_EDIT_VERSION_KEY = "tc_edit_version"; // do not change this string as it will break compatibility with other versions
export const tc_MIN_COMPATIBLE_VERSION_KEY = "tc_min_compatible_version"; // do not change this string as it will break compatibility with other versions
export const tc_MIN_VERSION_ERROR = "TC_MIN_VERSION_ERROR";

export const ORIGINAL_LANGUAGE = "originalLanguage";
export const TARGET_LANGUAGE = "targetLanguage";
export const TARGET_BIBLE = "targetBible";

export const toolCardCategories = {
  'kt': 'Key Terms',
  'names': 'Names',
  'other_terms': 'Other Terms',

  'culture': 'Culture',
  'figures': 'Figures of Speech',
  'numbers': 'Numbers',
  'discourse': 'Discourse',
  'grammar': 'Grammar',
  'other': 'Other'
};
