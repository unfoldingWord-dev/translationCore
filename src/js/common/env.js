// methods for accessing environment variables
// this is needed because many process.env values are no longer defined on the client side (looks to be
// security related).  To get the environment variable on client side can now use remote.process.env.

var isRunningClientSide = !process.env.HOME && !!window; // if environment not defined and we have a window
var processEnv = isRunningClientSide ? window.require('electron').remote.process.env : process.env;

/**
 * get Build number
 * @return {string}
 */
function getBuild() {
  return processEnv.BUILD;
}

/**
 * return appropriate process.env data
 * @return {*}
 */
function getEnv() {
  return processEnv;
}

var env = {
  getBuild,
  getEnv,
};

module.exports = env;
