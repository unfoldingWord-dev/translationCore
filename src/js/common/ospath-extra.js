// this is a patched ospath to work on client side in windows.  The problem was that ospath
// made use of process.env values which are no longer defined on the client side (looks to be
// security related).  To get the environment variable on client side can now use remote.process.env.
// TODO: electronite - investigate better solution
var path = require('path');
var os = require('os');

var processEnv = require('./env').getEnv();

function data () {
  switch (this.__platform || process.platform) {
    case 'win32': return path.resolve(processEnv.APPDATA);
    case 'darwin': return path.resolve(path.join(home.call(this), 'Library/Application Support/'));
    default: return processEnv.XDG_CONFIG_HOME
      ? path.resolve(processEnv.XDG_CONFIG_HOME)
      : path.resolve(path.join(home.call(this), '.config/'));
  }
}

function desktop () {
  return path.join(home.call(this), 'Desktop');
}

function home () {
  // io.js >= 2.3
  if ('homedir' in os) return os.homedir();

  switch (this.__platform || process.platform) {
    case 'win32': return path.resolve(processEnv.USERPROFILE);
    default: return path.resolve(processEnv.HOME)
  }
}

function tmp () {
  switch (this.__platform || process.platform) {
    case 'win32': return path.resolve(processEnv.TEMP);
    default: return path.resolve('/tmp');
  }
}

var ospath = {
  __platform: process.platform,
  data: data,
  desktop: desktop,
  home: home,
  tmp: tmp
};

module.exports = ospath;
