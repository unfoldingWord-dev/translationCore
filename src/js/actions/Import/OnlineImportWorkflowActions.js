import path from 'path-extra';
import fs from 'fs-extra';
// actions
import consts from '../../actions/ActionTypes';
import { deleteImportsFolder, deleteProjectFromImportsFolder } from '../../helpers/Import/ProjectImportFilesystemHelpers';
import * as AlertModalActions from '../../actions/AlertModalActions';
import * as OnlineModeConfirmActions from '../../actions/OnlineModeConfirmActions';
import * as ProjectImportStepperActions from '../ProjectImportStepperActions';
import * as MyProjectsActions from '../MyProjects/MyProjectsActions';
import * as ProjectDetailsActions from '../ProjectDetailsActions';
import * as ProjectInformationCheckActions from '../ProjectInformationCheckActions';
// helpers
import * as TargetLanguageHelpers from '../../helpers/TargetLanguageHelpers';
import {
  generateImportPath,
  verifyThisIsTCoreOrTStudioProject,
} from '../../helpers/Import/OnlineImportWorkflowHelpers';
import * as CopyrightCheckHelpers from '../../helpers/CopyrightCheckHelpers';
import {
  getProjectManifest,
  getProjectSaveLocation,
  getTranslate,
  getUsername,
} from '../../selectors';
import * as FileConversionHelpers from '../../helpers/FileConversionHelpers';
import * as ProjectDetailsHelpers from '../../helpers/ProjectDetailsHelpers';
import migrateProject from '../../helpers/ProjectMigration';
import Repo from '../../helpers/Repo';
import { isProjectSupported } from '../../helpers/ProjectValidation/ProjectStructureValidationHelpers';
import {
  openProject,
  closeProject,
  showInvalidVersionError,
} from '../MyProjects/ProjectLoadingActions';
import { delay } from '../../common/utils';
//constants
import { tc_MIN_VERSION_ERROR, IMPORTS_PATH } from '../../common/constants';
import * as ProjectImportFilesystemActions from './ProjectImportFilesystemActions';
import * as ProjectValidationActions from './ProjectValidationActions';

/**
 * Action that dispatches other actions to wrap up online importing
 */
export const onlineImport = () => (dispatch, getState) => new Promise((resolve, reject) => {
  const translate = getTranslate(getState());

  dispatch(OnlineModeConfirmActions.confirmOnlineAction(async () => {
    let importProjectPath = '';
    let link = '';

    try {
      await deleteImportsFolder();
      // Must allow online action before starting actions that access the internet
      link = getState().importOnlineReducer.importLink.trim();
      console.log('onlineImport() - link=' + link);

      dispatch(clearLink());
      // or at least we could pass in the locale key here.
      dispatch(AlertModalActions.openAlertDialog(translate('projects.importing_project_alert', { project_url: link }), true));

      const importPath = await generateImportPath(link);
      console.log('onlineImport() - import to: ' + importPath);

      await fs.ensureDir(importPath);
      console.log('onlineImport() - cloning repo into file system');
      await Repo.clone(link, importPath);
      const selectedProjectFilename = Repo.parseRemoteUrl(Repo.sanitizeRemoteUrl(link)).name;
      console.log('onlineImport() - selectedProjectFilename= ' + selectedProjectFilename);

      dispatch({ type: consts.UPDATE_SELECTED_PROJECT_FILENAME, selectedProjectFilename });
      importProjectPath = path.join(IMPORTS_PATH, selectedProjectFilename);

      // check if we can import the project
      console.log('onlineImport() - check if we can import the project');
      const isValid = verifyThisIsTCoreOrTStudioProject(importProjectPath);

      if (!isValid) {
        const errorMessage = translate('projects.online_import_error', { project_url: link, toPath: importProjectPath });
        console.warn('This is not a valid tStudio or tCore project we can migrate: ', errorMessage);
        throw errorMessage;
      }

      await isProjectSupported(importProjectPath, translate);
      const initialBibleDataFolderName = ProjectDetailsHelpers.getInitialBibleDataFolderName(selectedProjectFilename, importProjectPath);
      await migrateProject(importProjectPath, link, getUsername(getState()));
      // assign CC BY-SA license to projects imported from door43
      await CopyrightCheckHelpers.assignLicenseToOnlineImportedProject(importProjectPath);
      console.log('onlineImport() - start project validation');
      dispatch(ProjectValidationActions.initializeReducersForProjectImportValidation(false));
      await dispatch(ProjectValidationActions.validateProject(importProjectPath));
      const manifest = getProjectManifest(getState());
      const updatedImportPath = getProjectSaveLocation(getState());
      ProjectDetailsHelpers.fixBibleDataFolderName(manifest, initialBibleDataFolderName, updatedImportPath);

      if (!TargetLanguageHelpers.targetBibleExists(updatedImportPath, manifest)) {
        dispatch(AlertModalActions.openAlertDialog(translate('projects.loading_ellipsis'), true));
        console.log('onlineImport() - generate target bible');
        TargetLanguageHelpers.generateTargetBibleFromTstudioProjectPath(updatedImportPath, manifest);
        dispatch(ProjectInformationCheckActions.setSkipProjectNameCheckInProjectInformationCheckReducer(true));
        await delay(200);
        dispatch(AlertModalActions.closeAlertDialog());
        console.log('onlineImport() - validate project');
        await dispatch(ProjectValidationActions.validateProject(updatedImportPath));
      }

      const renamingResults = {};
      await dispatch(ProjectDetailsActions.updateProjectNameIfNecessary(renamingResults));
      const { projectDetailsReducer: { projectSaveLocation } } = getState();

      if (renamingResults.repoRenamed) {
        dispatch({ type: consts.UPDATE_SOURCE_PROJECT_PATH, sourceProjectPath: projectSaveLocation });
        dispatch({ type: consts.UPDATE_SELECTED_PROJECT_FILENAME, selectedProjectFilename: renamingResults.newRepoName });
        await delay(200);
      }
      await dispatch(ProjectImportFilesystemActions.move());

      if (renamingResults.repoRenamed) {
        await dispatch(ProjectDetailsActions.doRenamePrompting());
      }
      dispatch(MyProjectsActions.getMyProjects());

      // TODO: refactor this onlineImport method to remove project opening logic so we are not duplicating logic.

      const finalProjectPath = getProjectSaveLocation(getState());
      console.log('onlineImport() - project import complete: ' + finalProjectPath);
      await dispatch(openProject(path.basename(finalProjectPath), true));
      dispatch(AlertModalActions.closeAlertDialog());
      resolve();
    } catch (error) { // Catch all errors in nested functions above
      console.log('onlineImport() - import error:', error);
      const errorMessage = FileConversionHelpers.getSafeErrorMessage(error, translate('projects.online_import_error', { project_url: link, toPath: importProjectPath }));
      dispatch(recoverFailedOnlineImport(errorMessage));
      reject(errorMessage);
    }
  }));
});

/**
 * Performs recovery actions to cleanup after a failed online import
 * @param {string} errorMessage - A localized error message to show the user.
 * @returns {Function}
 */
export const recoverFailedOnlineImport = (errorMessage) => (dispatch) => {
  // TRICKY: clear last project first to avoid triggering autos-saving.
  dispatch(closeProject());

  if (errorMessage === tc_MIN_VERSION_ERROR) {
    dispatch(showInvalidVersionError());
  } else {
    dispatch(AlertModalActions.openAlertDialog(errorMessage));
  }
  dispatch(ProjectImportStepperActions.cancelProjectValidationStepper());
  dispatch({ type: 'LOADED_ONLINE_FAILED' });
  dispatch(deleteImportProjectForLink());
};

/**
 * TODO: this does not need to be an action.
 * @description - delete project (for link) from import folder
 */
export function deleteImportProjectForLink() {
  return ((dispatch, getState) => {
    const link = getState().importOnlineReducer.importLink;

    if (link) {
      const gitUrl = Repo.sanitizeRemoteUrl(link);
      let project = Repo.parseRemoteUrl(gitUrl);

      if (project) {
        deleteProjectFromImportsFolder(project.name);
      }
    }
  });
}

export function clearLink() {
  return {
    type: consts.IMPORT_LINK,
    importLink: '',
  };
}

export function getLink(importLink) {
  return {
    type: consts.IMPORT_LINK,
    importLink,
  };
}
