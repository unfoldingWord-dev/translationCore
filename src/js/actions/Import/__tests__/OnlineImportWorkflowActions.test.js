import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fs from 'fs-extra';
import path from 'path-extra';
// actions
import { recoverFailedOnlineImport } from '../OnlineImportWorkflowActions';
// helpers
import { IMPORTS_PATH, DCS_BASE_URL } from '../../../common/constants';
jest.mock('fs-extra');
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const importProjectName = 'es-419_tit_text_ulb';
const STANDARD_PROJECT = DCS_BASE_URL + '/royalsix/' + importProjectName + '.git';
const projectSaveLocation = IMPORTS_PATH;

// let mock_cloneManifest = null;

//mocking functions that are relevant to OnlineImportWorkflowActions but not required
// jest.mock('../js/helpers/Import/OnlineImportWorkflowHelpers', () => (
//   {
//     ...require.requireActual('../js/helpers/Import/OnlineImportWorkflowHelpers'),
//     clone: async () => {
//       return mock_saveJson(mock_cloneManifest);
//     }
//   }));

jest.mock('../../../helpers/ProjectMigration', () => jest.fn());
jest.mock('../ProjectValidationActions', () => (
  {
    ...require.requireActual('../ProjectValidationActions'),
    validateProject: () => ({ type: 'VALIDATE' }),
  }));
jest.mock('../ProjectImportFilesystemActions', () => ({
  deleteProjectFromImportsFolder: () => ({ type: 'DELETE_PROJECT_FROM_IMORTS' }),
  move: () => ((dispatch) => new Promise((resolve) => {
    dispatch({ type: 'MOVE' });
    resolve();
  })),
}));
jest.mock('../../MyProjects/MyProjectsActions', () => ({ getMyProjects: () => ({ type: 'GET_MY_PROJECTS' }) }));
jest.mock('../../MyProjects/ProjectLoadingActions', () => ({
  closeProject: () => ({ type: 'CLEAR_LAST_PROJECT' }),
  displayTools: jest.fn(() => ({ type: 'DISPLAY_TOOLS' })),
}));
jest.mock('../../../helpers/TargetLanguageHelpers', ()=> ({
  generateTargetBibleFromTstudioProjectPath: () => {},
  targetBibleExists:() => false,
}));
jest.mock('../../../helpers/ProjectValidation/ProjectStructureValidationHelpers', () => ({ ensureSupportedVersion: () => {} }));

describe('OnlineImportWorkflowActions.onlineImport', () => {
  let initialState = {};
  const manifest_ = {
    target_language: {
      id: 'es-419',
      name: 'es-419',
      direction: 'ltr',
    },
    generator: {
      name: 'ts-desktop',
      build: '132',
    },
    project: {
      id: 'tit',
      name: 'Titus',
    },
    resource: {
      id: 'ulb',
      name: 'unfoldingWord Literal Text',
    },
    tcInitialized: true,
    tc_version: 5,
  };

  beforeEach(() => {
    fs.__resetMockFS();
    // mock_cloneManifest = null;
    initialState = {
      importOnlineReducer: { importLink: STANDARD_PROJECT },
      settingsReducer: { onlineMode: true },
      projectDetailsReducer: {
        manifest: {},
        projectSaveLocation: projectSaveLocation,
      },
      localImportReducer: { selectedProjectFilename:'path' },
      loginReducer: { userdata: { userName: 'johndoe' } },
      projectInformationCheckReducer: {},
      projectValidationReducer: {},
    };
  });

  it('on import errors should call required actions', () => {
    const fileName = 'manifest.json';
    const cloneToPath = path.join(IMPORTS_PATH, importProjectName, fileName);
    fs.writeJSONSync(path.join(cloneToPath, 'manifest.json'), manifest_);

    const store = mockStore(initialState);
    store.dispatch(recoverFailedOnlineImport('import failed'));
    expect(store.getActions()).toMatchSnapshot();
  });
});
