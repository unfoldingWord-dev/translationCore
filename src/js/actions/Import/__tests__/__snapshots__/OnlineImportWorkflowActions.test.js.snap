// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OnlineImportWorkflowActions.onlineImport on import errors should call required actions 1`] = `
Array [
  Object {
    "type": "CLEAR_LAST_PROJECT",
  },
  Object {
    "alertMessage": "import failed",
    "loading": undefined,
    "type": "OPEN_ALERT_DIALOG",
  },
  Object {
    "showProjectValidationStepper": false,
    "type": "TOGGLE_PROJECT_VALIDATION_STEPPER",
  },
  Object {
    "type": "CLEAR_LAST_PROJECT",
  },
  Object {
    "type": "CLEAR_COPYRIGHT_CHECK_REDUCER",
  },
  Object {
    "type": "CLEAR_PROJECT_INFORMATION_REDUCER",
  },
  Object {
    "type": "CLEAR_MERGE_CONFLICTS_REDUCER",
  },
  Object {
    "type": "RESET_PROJECT_VALIDATION_REDUCER",
  },
  Object {
    "type": "GET_MY_PROJECTS",
  },
  Object {
    "type": "END_PROJECT_LOADING",
  },
  Object {
    "type": "LOADED_ONLINE_FAILED",
  },
]
`;
